share/dkg/dkg.go

```diff
@@ -0,0 +1,611 @@
+	// Index of the Verifier _issuing_ the ComplaintCommit
+	Index uint32
+	// DealerIndex being the index of the Dealer who issued the SecretCommits
+	DealerIndex uint32
+	// Deal that has been given from the Dealer (at DealerIndex) to this node
+	// (at Index)
+	Deal *vss.Deal
+	// Signature made by the verifier
+	Signature []byte
+}
+
+// ReconstructCommits holds the information given by a participants who reveals
+// the deal received from a peer that have received a ComplaintCommits.
+type ReconstructCommits struct {
+	// Index of the verifier who received the deal
+	Index uint32
+	// DealerIndex is the index of the dealer who issued the Deal
+	DealerIndex uint32
+	// Share contained in the Deal
+	Share *share.PriShare
+	// Signature over all over fields generated by the issuing verifier.
+	Signature []byte
+}
+
+type DistKeyGenerator struct {
+	suite abstract.Suite
+
+	index uint32
+	long  abstract.Scalar
+	pub   abstract.Point
+
+	participants []abstract.Point
+
+	t int
+
+	dealer    *vss.Dealer
+	verifiers map[uint32]*vss.Verifier
+
+	// list of commitments to each secret polynomials
+	commitments map[uint32]*share.PubPoly
+
+	// map of deals collected to reconstruct the full polynomial of a dealer
+	// key is index of the dealer. Once there is enough ReconstructCommits
+	// struct, the dkg will re-construct the polynomial and stores it into
+	// the list of commitments.
+	pendingReconstruct map[uint32][]*ReconstructCommits
+	reconstructed      map[uint32]bool
+}
+
+func NewDistKeyGenerator(suite abstract.Suite, longterm abstract.Scalar, participants []abstract.Point, r cipher.Stream, t int) (*DistKeyGenerator, error) {
+	d := new(DistKeyGenerator)
+	pub := suite.Point().Mul(nil, longterm)
+	// find our index
+	var found bool
+	for i, p := range participants {
+		if p.Equal(pub) {
+			found = true
+			d.index = uint32(i)
+			break
+		}
+	}
+	if !found {
+		return nil, errors.New("dkg: own public key not found in list of participants")
+	}
+	var err error
+	// generate our dealer / deal
+	ownSec := suite.Scalar().Pick(r)
+	d.dealer, err = vss.NewDealer(suite, longterm, ownSec, participants, r, t)
+	if err != nil {
+		return nil, err
+	}
+	// to receive the other deals
+	d.verifiers = make(map[uint32]*vss.Verifier)
+	d.commitments = make(map[uint32]*share.PubPoly)
+	d.pendingReconstruct = make(map[uint32][]*ReconstructCommits)
+	d.reconstructed = make(map[uint32]bool)
+	d.t = t
+	d.suite = suite
+	d.long = longterm
+	d.pub = pub
+	d.participants = participants
+	return d, nil
+}
+
+// DistDeals returns all the DistDeal that must be broadcasted to every
+// participants. The DistDeal corresponding to this DKG is already added
+// to this DKG and is ommitted from the returned map. To know
+// to which participant to give the DistDeal, loop over the keys as indexes in
+// the list of participants:
+//   for i,dd := range distDeals {
+//      sendTo(participants[i],dd)
+//   }
+// This method panics if it can't process it's own deal.
```

Comment:
>It's not as if the database has been corrupted or something else
In fact yes, if one can't process his own generated response that means something is wrong on the computer  memory, or in the math or whatever, but it's not a "casual" error. So I'm not sure what is best to do . I agree that panic-ing should be done as rarely as possible but at the same time, returning an error here seems to .. "undervalue" this error. well, maybe I'm overvalue-ing this error as well.. ^^